### DISCUSSION TITLE
How can I type annotate a factory helper's **kwargs to match that of the __init__ of the type being created?

### DISCUSSION THREAD

**Discussion by cjw296** [python/typing#1775] [Q&A]
So, for this code:

```python
class Collection:
    def make(self, type_: Type[T], **attrs) -> T:
        ...
```

How can I annotate `**attrs` such that a type checker, mypy in particular, will correctly ensure that any keywords passed are a valid subset of those that could be passed to `T`'s constructor?

**Comment by erictraut** [comment]
You can use a `ParamSpec`. There's no way to constrain it to only keyword parameters though. After all, the constructor may support positional parameters.

```python
from typing import Callable

class Collection:
    def make[**P, T](
        self, type_: Callable[P, T], *args: P.args, **kwargs: P.kwargs
    ) -> T: ...
```

Here's the same thing using older (pre-3.12) syntax:

```python
from typing import Callable, ParamSpec, TypeVar

P = ParamSpec("P")
T = TypeVar("T")

class Collection:
    def make(self, type_: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T: ...
```

**Reply by cjw296** [reply]
Okay, but notice the absence of `*args` in my original example...

Would this work?

```
from typing import Callable, ParamSpec, TypeVar

P = ParamSpec("P")
T = TypeVar("T")

class Collection:
    def make(self, type_: Callable[P, T], **kwargs: P.kwargs) -> T: ...
```

...but also, `type_` really should be a `Type[T]` not just anything that returns a `T`. How can I constrain or replace `Callable[P, T]` with `Type[T]` but still get the `P.kwargs`?

(also: wow, that 3.12 syntax is hard to read :-( - very sad to see Python is basically become C++, might as well just write in Rust and at least get performance for the price of hard-to-read code!)

**Reply by Daverball** [reply]
No, `ParamSpec` always requires you to use both `*args` and `**kwargs` when unpacking, since otherwise the upper bound for the `ParamSpec` would have to implicitly change based on how you unpack it and that would then raise questions about what happens when that unpack happens several layers deep or you unpack the same `ParamSpec` two different ways. So for simplicity you are not allowed to omit `*args`.

This is not set in stone and could later be changed with a follow-up PEP that specifies explicit upper bounds for `ParamSpec` (there's currently no way to set an upper bound for a `ParamSpec` like you would be able to for a `TypeVar`). I think it's easier to justify allowing people to omit one or the other when the upper bound has been eagerly set to only contain only one or the other, although I would suspect that this restriction would still stay to keep things as simple as possible for type checkers.

You can have only `*args` using `TypeVarTuple` instead of `ParamSpec`, but there's currently no equivalent for mapping based variadic generics, so you can't have only `**kwargs`.

Also `type[T]` is a subtype of `Callable[P, T]` where `P` represents the arguments of the constructor of `T`. I don't see a reason to arbitrarily restrict this function to `type[T]`, unless you want to be able to use it in `isinstance` calls or do something else where you need an instance of type.

The short answer is, you currently can't do what you want to be able to do here. Once intersections become a thing you could probably write this as `type[T] & Callable[P, T]`. Until then, Eric's solution is the best you're going to get. You can rely on runtime assertions for the rest in the meantime, i.e. start your function with `assert isinstance(type_, type)`, and even `assert not args` if you absolutely don't want to allow them for whatever reason.

**Reply by cjw296** [reply]
> You can have only *args using TypeVarTuple instead of ParamSpec, but there's currently no equivalent for mapping based variadic generics, so you can't have only **kwargs.

Well that sucks :-( I would have thought "only accepting kwargs" is a more common pattern over Python's history, so it's disappointing _that's_ the option that's not supported :-/ 

> Also type[T] is a subtype of Callable[P, T] where P represents the arguments of the constructor of T. I don't see a reason to arbitrarily restrict this function to type[T], unless you want to be able to use it in isinstance calls or do something else where you need an instance of type.

Yes, exactly that, this package needs a type, not some arbitrary callable. Once again python typing ends up burning cpu at runtime.

Pragmatically, I either leave it as:

```python
from typing import Type, Any, TypeVar

T = TypeVar("T")

class Collection:
    def make(self, type_: Type[T], **kwargs: Any) -> T: ...
```

...or I have to:

```python

from typing import Callable, ParamSpec, TypeVar

P = ParamSpec("P")
T = TypeVar("T")

class Collection:
    def make(self, type_: Callable[P, T], *args, P.args, **kwargs: P.kwargs) -> T:
         assert not args
         assert isinstance(type_, type)
         ...
```

*sigh*

**Reply by bzoracler** [reply]
Here's a contortion which works (practically) on mypy ([Playground example](https://mypy-play.net/?mypy=latest&python=3.12&gist=b1eaf7ea62bda8a57bd398b4c672265f)):

```python
from collections.abc import Callable
from typing_extensions import Any, Concatenate, ParamSpec, Protocol, TypeAlias, TypeVar


Self: TypeAlias = Any
T = TypeVar("T")
P = ParamSpec("P")


class KWArgsOnly(Protocol):
    def __call__(self, factory: type[Any], /, **kwargs: object) -> Any: ...


def reject_non_type_and_non_kwargs(
    make: Callable[Concatenate[Self, Callable[P, T], P], T],
) -> Callable[Concatenate[Callable[P, T], P], T] | KWArgsOnly:
    return make  # type: ignore[return-value]


class Collection:
    @reject_non_type_and_non_kwargs
    def make(self, factory: Callable[P, T], /, *args: P.args, **kwargs: P.kwargs) -> T:
        return factory(*args, **kwargs)


class Instance:
    def __init__(self, a: int, b: str, c: list[int]) -> None: ...


def non_type_factory(a: int, b: str, c: list[int]) -> Instance:
    return Instance(a, b, c)
```

```python
reveal_type(Collection().make(Instance, a=1, b="", c=[]))  # OK
Collection().make(Instance, 1, b="", c=[])  # Error: passed positional argument
Collection().make(non_type_factory, a=1, b="", c=[])  # Error: passed non-type
```

The idea is that `@reject_non_type_and_non_kwargs` returns the `make` method's original signature unioned with a callback that only accepts a `type` and keyword arguments. This means that you'll get errors at the call site if you try to pass `non_type_factory` (a function) and/or positional arguments afterwards.

This doesn't protect you from making a bad signature for `make` though, so ensure that you have properly demarcated your positional prefix arguments from the constructor.

**Reply by Daverball** [reply]
That's a neat trick, however it changes the return type to `Instance | Any`, here's a slightly adjusted version that properly preserves the return type: https://mypy-play.net/?mypy=latest&python=3.12&gist=63473b4b5ef209952fa5ea5c1f2ba7d5

**Reply by cjw296** [reply]
It really does sound like there's a deficiency in Python's typing, where `ParamSpec` should be usable to say, *either* `args` or `kwargs` or a way of saying `ParamSpec(...partial=True)` or some such. Has that discussion been has? If so, where?

If not, where should it be had?

The fact that `partial` is still typed as `Any` in typeshed seems bad to me :-(

**Reply by Daverball** [reply]
The Python Discourse would be the current best place for typing related discussions: https://discuss.python.org/

`ParamSpec` adjacent topics and discussions have come up several times. Python function signatures are very versatile, but that also makes them very difficult to deal with in static typing contexts. One big issue is expressiveness, since we currently have no way to spell a complex signature in a pure type expression. We instead have to rely on callback protocols, which only gives us part of the functionality we need and some things are only very difficult or awkward to express. This is a relative recent discussion about a possible syntax extension to fix that part of the problem: https://discuss.python.org/t/reviving-the-hybrid-keyword-arrow-syntax-for-callables/52499

Once we can actually express all kinds of signatures in type expressions, it will be much easier to add the `bound` argument to `ParamSpec`, which would allow your use-case (although you'd still have to write `*args: P.args, **kwargs: P.kwargs`). `TypeVarMapping` has been discussed several times as well, but after how difficult `ParamSpec` and `TypeVarTuple` turned out to implement and how long it took for even just basic support to land in more than one type checker, the brakes were hit on adding more kinds of type parameters.

The main focus for Python typing currently is to write a complete typing spec and compliance test suite, so type checkers have to waste less time individually figuring out how certain features are supposed to work, because the wording in the original PEP was too loose/ambiguous.

**Reply by cjw296** [reply]
Has any progress been made on this?

> (although you'd still have to write `*args: P.args, **kwargs: P.kwargs`).

Not sure what you mean by this? My requirement is to be able to explicitly say that only `kwargs` from `P` are accepted.

**Reply by Daverball** [reply]
No progress has been made on this, no. What I mean by that is that the restriction of "no positional arguments" would be handled through the `bound` parameter of the `ParamSpec`, but that wouldn't necessarily change the requirement that you always need to unpack both components of `P`.

So the type checker would ensure that no positional arguments can be passed in, but your generic signature would still need to include them anyways (unless that restriction is lifted for upper bounds that ensure that either one of the components is definitely always empty).

A `TypeVarMapping` might be an alternative avenue, although given that `TypeVarTuple` and `ParamSpec` still aren't fully supported in pytype to this day, it would probably be met with little enthusiasm, by the people that would have to implement that feature in their type checkers.

**Reply by cjw296** [reply]
> but that wouldn't necessarily change the requirement that you always need to unpack both components of P.

Sorry, not following this, could you provide a code example of what you mean?

> So the type checker would ensure that no positional arguments can be passed in, but your generic signature would still need to include them anyways

This sounds pretty sub-optimal :-/ 

What's pytype? That's not a phrase I've heard before...

**Reply by Daverball** [reply]
> Sorry, not following this, could you provide a code example of what you mean?

The spec for `ParamSpec` currently states that if `P.args` is used in a function's signature `P.kwargs` must also be used and vice versa,  there also can't be any arguments in between, the only thing you're allowed to do is prepend arguments, like you would with `Concatenate`. This is to ensure that no matter what `P` solves to, the call is still possible. Otherwise the same P would need to check every single place it's used in order to ensure consistency.

```python
def allowed[**P](self, *args: P.args, **kwargs: P.kwargs): ...
def positional_only_prepend_allowed[**P](self, arg: int, /, *args: P.args, **kwargs: P.kwargs): ...
def no_kwargs_disallowed[**P](self, *args: P.args): ...
def no_args_disallowed[**P](self, **kwargs: P.kwargs): ...
def middle_extra_kwarg_disallowed[**P](self, *args: P.args, my_new_kwarg: int = 0, **kwargs: P.kwargs): ...
```

With upper bounds on `P` this restriction could potentially be relaxed in some cases, e.g. with an upper bound of `def (**kwargs: Any)` you may be allowed to omit `P.args` and define an arbitrary number of preceding positional only arguments. Although the required implementation complexity makes this probably almost certainly not worth it.

> What's pytype? That's not a phrase I've heard before...

It's a Python type checker made by Google. It performs very deep type inference and allows a lot of things other type checkers would  disallow because their inference is too shallow to ensure type safety without an explicit type annotation. So it can be useful in code that doesn't make heavy use of type annotations, since it will still be able to detect a lot of problems anyways.

**Reply by cjw296** [reply]
The irony's not lost on me that some of the things that have made Python so powerful over the last two or three decades end up being the things that are hardest to express in static typing...

Thanks for the pytype pointer; is this one, right? https://github.com/google/pytype

**Reply by Daverball** [reply]
Correct

**Reply by cjw296** [reply]
Hmm, unfortunately, pytype seems to trip up on [much simpler stuff](https://github.com/google/pytype/issues/1867) that mypy is happy with :-/

**Comment by cjw296** [comment]
https://github.com/python/typing/discussions/1500 seems to be the same ballpark as this.

**Reply by Daverball** [reply]
That sounds more like a proxy type to me, which has been requested many times before, e.g. https://github.com/python/typing/issues/802 but we don't have anything like that yet.

There definitely are some open questions for proxy types, like should a proxy be allowed to stand-in for the real type, what about partial proxies that don't proxy everything etc. It would definitely need a PEP to explore and answer some of those questions, so we have a concrete proposal that can be implemented in a type checker as a proof of concept.
